contexts/currency-context.tsx

```diff
@@ -0,0 +1,211 @@
+import React, { useCallback } from 'react'
+import { useAccount, useQuery, useQueryClient } from 'wagmi'
+import { readContracts } from '@wagmi/core'
+
+import { fetchAssets } from '../api/token'
+import { CONTRACT_ADDRESSES } from '../utils/addresses'
+import { CouponOracle__factory, IERC20__factory } from '../typechain'
+import { Currency } from '../model/currency'
+import { DecimalNumber } from '../utils/decimal-number'
+import { Asset } from '../model/asset'
+
+type CurrencyContext = {
+  assets: Asset[]
+  balances: { [key in `0x${string}`]: DecimalNumber }
+  // contract address => token address => allowance
+  allowances: {
+    [key in `0x${string}`]: { [key in `0x${string}`]: DecimalNumber }
+  }
+  prices: { [key in `0x${string}`]: DecimalNumber }
+  invalidateBalances: () => void
+  invalidateAllowances: () => void
+}
+
+const Context = React.createContext<CurrencyContext>({
+  assets: [],
+  balances: {},
+  allowances: {},
+  prices: {},
+  invalidateBalances: () => {},
+  invalidateAllowances: () => {},
+})
+
+export const CurrencyProvider = ({ children }: React.PropsWithChildren<{}>) => {
+  const queryClient = useQueryClient()
+
+  const { address: userAddress } = useAccount()
+
+  const { data: assets } = useQuery(
+    ['assets'],
+    async () => {
+      return fetchAssets()
+    },
+    {
+      refetchInterval: 1000,
+      refetchIntervalInBackground: true,
+    },
+  )
+
+  const fetchCurrencies = async () => {
+    const assets = await fetchAssets()
+    return Object.values<Currency>(
+      assets
+        .reduce(
+          (acc: Currency[], asset) => [
+            ...acc,
+            asset.underlying,
+            ...asset.collaterals,
+          ],
+          [],
+        )
+        .reduce((acc: { [key in `0x${string}`]: Currency }, currency) => {
+          return {
+            ...acc,
+            [currency.address]: currency,
+          }
+        }, {}),
+    )
+  }
+
+  const { data: prices } = useQuery(
```

Comment:
How about this?

```suggestion
  const { data: pricesMap } = useQuery(
```

Reply:
I think `prices` is good enough since typescript types already imply that it's a map.